### PR TITLE
EDU-764 Create backend for comments

### PR DESCRIPTION
**Pull request by adrianaluchian**
Closes https://educandu.atlassian.net/browse/EDU-764

### CODE REVIEW

File: src/domain/permissions.js [hunk 1]
```diff
@@ -115,5 +119,7 @@ export default {
   OWN_ROOMS,
   DELETE_FOREIGN_ROOMS,
   AUTORIZE_ROOMS_RESOURCES,
-  JOIN_PRIVATE_ROOMS
+  JOIN_ROOMS,
+  CREATE_DOCUMENT_COMMENT,
+  MANAGE_DOCUMENT_COMMENT
```
Comment (ahelmberger) [inline src/domain/permissions.js:124]: In combination with db "entitites" we mostly used the plural so far (like "manage-storage-plans" etc.). I'm not insisting, just preferring :-)

File: src/domain/schemas/comment-schemas.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+export const commentDBSchema = joi.object({
+  _id: idOrKeySchema.required(),
+  documentId: idOrKeySchema.required(),
+  createdOn: joi.date().required(),
+  createdBy: idOrKeySchema.required(),
+  deletedOn: joi.date().allow(null),
+  deletedBy: idOrKeySchema.allow(null),
+  topic: joi.string().max(maxCommentTopicLength).required(),
+  text: joi.string().max(maxCommentTextLength).allow(null)
```
Comment (ahelmberger) [inline src/domain/schemas/comment-schemas.js:27]: I think this is another one of the examples where you like `null` and I don't (as the user can hardly ever enter it), but we can postpone that for an eventual streamlining session.

File: src/server/comment-controller.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      '/api/v1/comments',
+      jsonParser,
+      needsPermission(permissions.CREATE_DOCUMENT_COMMENT),
+      validateBody(putCommentBodySchema),
+      (req, res) => this.handlePutComment(req, res)
+    );
+
+    router.post(
+      '/api/v1/comments/:commentId',
+      [needsPermission(permissions.MANAGE_DOCUMENT_COMMENT), jsonParser, validateBody(postCommentBodySchema)],
```
Comment (ahelmberger) [inline src/server/comment-controller.js:95]: Little thing: we should also validate the `:commentId` param

File: src/server/comment-controller.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  async handlePutComment(req, res) {
+    const { user } = req;
+    const data = req.body;
+
+    const document = await this.documentService.getDocumentById(data.documentId);
+
+    if (!document) {
+      throw new BadRequest(`Unknown document id '${data.documentId}'`);
+    }
+
```
Comment (ahelmberger) [inline src/server/comment-controller.js:42]: Maybe we could check here also if the document is an internal one and throw if it isn't. Just to make sure...?

File: src/domain/permissions.js (hunk 1 above)
Comment (adrianaluchian) [reply src/domain/permissions.js]: Nicely spotting, I of course aligned the namings, thanks!

File: src/domain/schemas/comment-schemas.js (hunk 2 above)
Comment (adrianaluchian) [reply src/domain/schemas/comment-schemas.js]: I agree with the empty string default, changed!

File: src/server/comment-controller.js (hunk 3 above)
Comment (adrianaluchian) [reply src/server/comment-controller.js]: Indeed, added validation.

Comment (ahelmberger) [reply src/server/comment-controller.js]: :+1:

**Review by ahelmberger — APPROVED** [review]

**Pull request merged**